File: CFSharp/CFSharp.C/Tests/MonadPoliceTests.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using NSubstitute;
using NSubstitute.Experimental;
using NUnit.Framework;

namespace CFSharp.C.Tests
{
    public class MonadPoliceTests
    {
        private IEmailGateway _emailer;
        private ImDave _dave;
        private MonadPolice _subject;

        [SetUp]
        public void SetUp()
        {
            _emailer = Substitute.For<IEmailGateway>();
            _dave = Substitute.For<ImDave>();
            _subject = new MonadPolice(_dave, _emailer);
        }

        [Test]
        public void ShouldNotifyXerxWithEveryMBomb()
        {
            var ramblings = new [] {
                "tea?",
                "couldn't we use a monad for that?",
                "anyone for tea?",
                "blah blah blah haskell blah blah monad blah blah",
                "a cup of tea is a lot like the Maybe monad..."
            };
            _dave.RecentRamblings().Returns(ramblings);

            _subject.Surveil();

            _emailer.Received().Send("xerxesb", Arg.Any<string>(), Arg.Any<string>(), ramblings[1]);
            _emailer.Received().Send("xerxesb", Arg.Any<string>(), Arg.Any<string>(), ramblings[3]);
            _emailer.Received().Send("xerxesb", Arg.Any<string>(), Arg.Any<string>(), ramblings[4]);
            _emailer.ReceivedWithAnyArgs(3).Send(null, null, null, null);
        }

        [Test]
        public void Sample()
        {
            var ramblings = new [] {
                "tea?",
                "couldn't we use a monad for that?",
                "anyone for tea?",
                "blah blah blah haskell blah blah monad blah blah",
                "a cup of tea is a lot like the Maybe monad..."
            };
            _dave.RecentRamblings().Returns(ramblings);
            var emailer = new ConsoleEmailGateway();

            new MonadPolice(_dave, emailer).Surveil();
        }

        private class ConsoleEmailGateway : IEmailGateway
        {
            public void Send(string to, string @from, string subject, string message)
            {
                Console.WriteLine(subject + ": " + message);
            }
        }
    }
}

File: CFSharp/CFSharp.F/MonadPoliceTests.fs
module CFSharp.F.MonadPoliceTests

open NUnit.Framework
open NSubstitute
open FsUnit
open MonadPolice
open TestHelper

type MonadPoliceTests() =
    [<DefaultValue>] val mutable _emailer : IEmailGateway
    [<DefaultValue>] val mutable _dave : ImDave
    [<DefaultValue>] val mutable _subject : MonadPolice

    [<SetUp>]
    member this.SetUp() =
        this._emailer <- Substitute.For<IEmailGateway>()
        this._dave <- Substitute.For<ImDave>()
        this._subject <- MonadPolice(this._dave, this._emailer)

    [<Test>]
    member this.ShouldNotifyXerxWithEveryMBomb() =
        let ramblings =
            [ "tea?"
            ; "couldn't we use a monad for that?"
            ; "anyone for tea?"
            ; "blah blah blah haskell blah blah monad blah blah"
            ; "a cup of tea is a lot like the Maybe monad..." 
            ]
        this._dave.RecentRamblings() |> Returns (Seq.ofList ramblings)

        this._subject.Surveil()

        (this._emailer |> Received).Send "xerxesb" (Any<string>()) (Any<string>()) ramblings.[1]
        (this._emailer |> Received).Send "xerxesb" (Any<string>()) (Any<string>()) ramblings.[3]
        (this._emailer |> Received).Send "xerxesb" (Any<string>()) (Any<string>()) ramblings.[4]
        (this._emailer |> ReceivedWithAnyArgs 3).Send (Any<string>()) (Any<string>()) (Any<string>()) (Any<string>())
